waspc/src/Wasp/Generator/ServerGenerator/EmailG.hs

```diff
@@ -0,0 +1,143 @@
+import Wasp.AppSpec.Valid (getApp)
+import Wasp.Generator.FileDraft (FileDraft)
+import Wasp.Generator.Monad (Generator)
+import qualified Wasp.Generator.ServerGenerator.Common as C
+import qualified Wasp.SemanticVersion as SV
+import Wasp.Util ((<++>))
+
+genEmail :: AppSpec -> Generator [FileDraft]
+genEmail spec = case maybeEmailSender of
+  Just emailSender ->
+    sequence
+      [ genIndex emailSender
+      ]
+      <++> genCore emailSender
+  Nothing -> return []
+  where
+    maybeEmailSender = AS.App.emailSender $ snd $ getApp spec
+
+genIndex :: EmailSender -> Generator FileDraft
+genIndex email = return $ C.mkTmplFdWithDstAndData srcPath dstPath (Just tmplData)
+  where
+    srcPath = C.srcDirInServerTemplatesDir </> [relfile|email/index.ts|]
+    dstPath = C.serverSrcDirInServerRootDir </> [relfile|email/index.ts|]
+    tmplData = getEmailProvidersJson email
+
+genCore :: EmailSender -> Generator [FileDraft]
+genCore email =
+  sequence
+    [ genCoreIndex email,
+      copyTmplFile [relfile|email/core/types.ts|],
+      genCoreHelpers email
+    ]
+    <++> genSmtp email
+    <++> genSendGrid email
+    <++> genMailgun email
+
+genCoreIndex :: EmailSender -> Generator FileDraft
+genCoreIndex email = return $ C.mkTmplFdWithDstAndData srcPath dstPath (Just tmplData)
+  where
+    srcPath = C.srcDirInServerTemplatesDir </> [relfile|email/core/index.ts|]
+    dstPath = C.serverSrcDirInServerRootDir </> [relfile|email/core/index.ts|]
+    tmplData = getEmailProvidersJson email
+
+genCoreHelpers :: EmailSender -> Generator FileDraft
+genCoreHelpers email = return $ C.mkTmplFdWithDstAndData srcPath dstPath (Just tmplData)
+  where
+    srcPath = C.srcDirInServerTemplatesDir </> [relfile|email/core/helpers.ts|]
+    dstPath = C.serverSrcDirInServerRootDir </> [relfile|email/core/helpers.ts|]
+    tmplData =
+      object
+        [ "senderDefaults"
+            .= object
+              [ "email" .= AS.EmailSender.email emailFrom,
+                "title" .= title,
+                "isTitleDefined" .= isJust title
+              ]
+        ]
+    emailFrom = AS.EmailSender.defaultFrom email
+    title = AS.EmailSender.title emailFrom
+
+genSmtp :: EmailSender -> Generator [FileDraft]
+genSmtp email =
+  case AS.EmailSender.provider email of
+    AS.EmailSender.SMTP ->
+      sequence
+        [ copyTmplFile [relfile|email/core/providers/smtp.ts|]
+        ]
+    _ -> return []
+
+genSendGrid :: EmailSender -> Generator [FileDraft]
+genSendGrid email =
+  case AS.EmailSender.provider email of
+    AS.EmailSender.SendGrid ->
+      sequence
+        [ copyTmplFile [relfile|email/core/providers/sendgrid.ts|]
+        ]
+    _ -> return []
+
+genMailgun :: EmailSender -> Generator [FileDraft]
+genMailgun email =
+  case AS.EmailSender.provider email of
+    AS.EmailSender.Mailgun ->
+      sequence
+        [ copyTmplFile [relfile|email/core/providers/mailgun.ts|]
+        ]
+    _ -> return []
```

Comment:
Could you unite all these into one function called `genEmailSender` (I made this name up, you might have a better idea of what is a good name here, I had a bit of trouble figuring it out from the context) and then it would have one big `case` which would check which provider it is and based on it return correct filedraft?

You could reuse than the `copyTmplFile [relfile|email/core/providers/` parts also I believe.

Reply:
That's very nice idea, I'll try to rewrite this piece to have all of the provider data grouped in a place and then we just switch between them.